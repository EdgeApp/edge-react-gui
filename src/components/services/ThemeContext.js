// @flow

import { cacheStyles, makeThemeContext } from 'react-native-patina'

import { womanTheme } from '../../theme/variables/womanTheme'
import { type Theme } from '../../types/Theme.js'

export type { Theme }
export { cacheStyles }

/**
 * Utility type for declaring `withTheme` components.
 */
export type ThemeProps = {
  theme: Theme
}

// Provide the theme context methods:
const themeContext = makeThemeContext(womanTheme)
export const { ThemeProvider, useTheme, withTheme, changeTheme, getTheme, watchTheme } = themeContext
