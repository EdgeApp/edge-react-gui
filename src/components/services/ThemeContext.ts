import { cacheStyles, makeThemeContext } from 'react-native-patina'

import { config } from '../../theme/appConfig'
import { Theme } from '../../types/Theme'

export type { Theme }
export { cacheStyles }

/**
 * Utility for declaring `withTheme` components.
 */
export type ThemeProps = {
  theme: Theme
}

// Provide the theme context methods:
const themeContext = makeThemeContext(config.darkTheme)
export const { ThemeProvider, useTheme, withTheme, changeTheme, getTheme, watchTheme } = themeContext
