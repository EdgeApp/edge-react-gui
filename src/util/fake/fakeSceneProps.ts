import { AppParamList, EdgeSceneProps, NavigationProp } from '../../types/routerTypes'

export const fakeNavigation: NavigationProp<any> = {
  addListener() {
    return () => undefined
  },
  removeListener() {},
  isFocused() {
    return true
  },

  navigate() {},
  push() {},
  replace() {},
  setParams(params) {},

  goBack() {},
  pop() {},
  popToTop() {},

  dispatch() {},
  reset() {},
  canGoBack() {
    return true
  },
  getId() {
    return undefined
  },
  getState() {
    return {
      key: 'foo',
      index: 0,
      routes: [],
      routeNames: [],
      type: 'bar',
      stale: false
    }
  },
  getParent() {
    throw new Error('not implemented')
  },
  setOptions() {}
}

export function fakeSceneProps<Name extends keyof AppParamList>(name: Name, params: AppParamList[Name]): EdgeSceneProps<Name> {
  return {
    navigation: fakeNavigation,
    route: {
      name: name,
      key: name + '-0',
      params
    } as any
  }
}
