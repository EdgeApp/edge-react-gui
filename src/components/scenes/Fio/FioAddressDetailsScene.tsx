import { EdgeCurrencyWallet } from 'edge-core-js'
import * as React from 'react'
import { Alert } from 'react-native'
import IonIcon from 'react-native-vector-icons/Ionicons'

import { lstrings } from '../../../locales/strings'
import { connect } from '../../../types/reactRedux'
import { EdgeSceneProps } from '../../../types/routerTypes'
import { BUNDLED_TXS_AMOUNT_ALERT, findWalletByFioAddress } from '../../../util/FioAddressUtils'
import { SceneWrapper } from '../../common/SceneWrapper'
import { ConnectWalletsConnector as ConnectWallets } from '../../FioAddress/ConnectWallets'
import { showError } from '../../services/AirshipInstance'
import { cacheStyles, Theme, ThemeProps, withTheme } from '../../services/ThemeContext'
import { SettingsHeaderRow } from '../../settings/SettingsHeaderRow'
import { SettingsTappableRow } from '../../settings/SettingsTappableRow'
import { EdgeText } from '../../themed/EdgeText'

interface StateProps {
  fioWallets: EdgeCurrencyWallet[]
}

interface OwnProps extends EdgeSceneProps<'fioAddressDetails'> {}

interface LocalState {
  fioWalletLoading: boolean
  fioWallet: EdgeCurrencyWallet | null
}

type Props = StateProps & ThemeProps & OwnProps

export class FioAddressDetails extends React.Component<Props, LocalState> {
  state: LocalState = {
    fioWalletLoading: false,
    fioWallet: null
  }

  componentDidMount() {
    const { route } = this.props
    const { fioAddressName } = route.params
    if (!fioAddressName) {
      Alert.alert(lstrings.fio_address_details_screen_alert_title, lstrings.fio_address_details_screen_alert_message, [
        { text: lstrings.fio_address_details_screen_alert_button }
      ])
    }
    this.findFioWallet().catch(err => showError(err))
  }

  findFioWallet = async () => {
    const { fioWallets, route } = this.props
    const { fioAddressName } = route.params

    this.setState({ fioWalletLoading: true })
    const fioWallet = await findWalletByFioAddress(fioWallets, fioAddressName)
    this.setState({ fioWalletLoading: false, fioWallet })
  }

  _onPressAccountSettings = (): void => {
    const { navigation, route } = this.props
    const { fioAddressName, bundledTxs } = route.params
    const { fioWallet } = this.state
    if (fioWallet) {
      navigation.navigate('fioAddressSettings', {
        fioWallet,
        fioAddressName,
        bundledTxs,
        refreshAfterAddBundledTxs: true
      })
    } else {
      showError(lstrings.fio_wallet_missing_for_fio_address)
    }
  }

  renderAccountSettings = () => {
    const {
      theme,
      route: {
        params: { bundledTxs }
      }
    } = this.props
    const styles = getStyles(theme)

    if (bundledTxs < BUNDLED_TXS_AMOUNT_ALERT) {
      return (
        <SettingsTappableRow onPress={this._onPressAccountSettings}>
          <IonIcon name="ios-warning" color={theme.warningIcon} style={styles.settingsIcon} />
          <EdgeText style={styles.settingsWarning} numberOfLines={4}>
            {!bundledTxs ? lstrings.fio_address_details_no_bundled_txs : lstrings.fio_address_details_bundled_txs_out_soon}
          </EdgeText>
        </SettingsTappableRow>
      )
    }

    return (
      <SettingsTappableRow label={lstrings.fio_address_details_screen_manage_account_settings} onPress={this._onPressAccountSettings}>
        <IonIcon name="ios-settings" color={theme.icon} style={styles.settingsIcon} />
      </SettingsTappableRow>
    )
  }

  render() {
    const { navigation, theme, route } = this.props
    const { fioAddressName, bundledTxs } = route.params
    const styles = getStyles(theme)
    const bundledTxsLabel = `${lstrings.fio_address_details_screen_bundled_txs}: ${bundledTxs}`

    return (
      <SceneWrapper background="theme">
        <EdgeText style={styles.bundledTxs}>{bundledTxsLabel}</EdgeText>
        {this.renderAccountSettings()}
        <SettingsHeaderRow
          icon={<IonIcon name="ios-link" color={theme.primaryText} size={theme.rem(1.5)} />}
          label={lstrings.fio_address_details_connect_to_wallets}
        />
        <ConnectWallets fioAddressName={fioAddressName} fioWallet={this.state.fioWallet} navigation={navigation} disabled={this.state.fioWalletLoading} />
      </SceneWrapper>
    )
  }
}

const getStyles = cacheStyles((theme: Theme) => ({
  text: {
    color: theme.primaryText,
    fontSize: theme.rem(1)
  },
  bundledTxs: {
    fontSize: theme.rem(0.75),
    color: theme.primaryText,
    textAlign: 'center',
    marginTop: theme.rem(-0.5),
    paddingBottom: theme.rem(0.75)
  },
  settingsIcon: {
    fontSize: theme.rem(1.5),
    paddingHorizontal: theme.rem(0.5)
  },
  settingsWarning: {
    fontSize: theme.rem(0.75),
    color: theme.warningText,
    paddingLeft: theme.rem(0.5),
    width: '75%'
  }
}))

export const FioAddressDetailsScene = connect<StateProps, {}, OwnProps>(
  state => ({
    fioWallets: state.ui.wallets.fioWallets
  }),
  dispatch => ({})
)(withTheme(FioAddressDetails))
