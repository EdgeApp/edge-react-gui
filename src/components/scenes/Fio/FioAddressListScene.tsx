import { EdgeCurrencyWallet } from 'edge-core-js'
import * as React from 'react'
import { ActivityIndicator, Image, ScrollView, View } from 'react-native'
import IonIcon from 'react-native-vector-icons/Ionicons'

import { refreshAllFioAddresses } from '../../../actions/FioAddressActions'
import fioAddressLogo from '../../../assets/images/fio/fio_logo.png'
import { Fontello } from '../../../assets/vector'
import { lstrings } from '../../../locales/strings'
import { connect } from '../../../types/reactRedux'
import { EdgeSceneProps } from '../../../types/routerTypes'
import { FioAddress, FioDomain } from '../../../types/types'
import { SceneWrapper } from '../../common/SceneWrapper'
import { FioNameRow } from '../../FioAddress/FioName'
import { FullScreenLoader } from '../../progress-indicators/FullScreenLoader'
import { showError } from '../../services/AirshipInstance'
import { cacheStyles, Theme, ThemeProps, withTheme } from '../../services/ThemeContext'
import { ClickableText } from '../../themed/ClickableText'
import { EdgeText } from '../../themed/EdgeText'
import { Fade } from '../../themed/Fade'
import { SceneHeader } from '../../themed/SceneHeader'

interface LocalState {
  initLoading: boolean
  prevLoading: boolean
}

interface StateProps {
  fioAddresses: FioAddress[]
  fioDomains: FioDomain[]
  fioWallets: EdgeCurrencyWallet[]
  loading: boolean
  isConnected: boolean
}

interface DispatchProps {
  refreshAllFioAddresses: () => Promise<void>
}

interface OwnProps extends EdgeSceneProps<'fioAddressList'> {}

type Props = StateProps & DispatchProps & OwnProps & ThemeProps

export class FioAddressList extends React.Component<Props, LocalState> {
  willFocusSubscription: (() => void) | null = null
  state: LocalState = {
    initLoading: true,
    prevLoading: false
  }

  static getDerivedStateFromProps(props: Props, state: LocalState): LocalState | null {
    const { loading } = props
    const { prevLoading, initLoading } = state
    if (!loading && prevLoading && initLoading) {
      return {
        prevLoading: loading,
        initLoading: false
      }
    }
    if (loading !== prevLoading) {
      return {
        initLoading,
        prevLoading: loading
      }
    }

    return null
  }

  async fetchData() {
    const { refreshAllFioAddresses, isConnected } = this.props
    if (!isConnected) {
      showError(lstrings.fio_network_alert_text)
    }
    await refreshAllFioAddresses()
  }

  componentDidMount(): void {
    this.willFocusSubscription = this.props.navigation.addListener('focus', () => {
      this.fetchData().catch(err => showError(err))
    })
  }

  componentWillUnmount(): void {
    if (this.willFocusSubscription != null) this.willFocusSubscription()
  }

  onAddressPress = (fioAddress: FioAddress) => {
    const { navigation } = this.props
    const { name, bundledTxs } = fioAddress
    navigation.navigate('fioAddressDetails', {
      fioAddressName: name,
      bundledTxs
    })
  }

  onDomainPress = (fioDomain: FioDomain) => {
    const { fioWallets, navigation } = this.props
    const { name, expiration, walletId, isPublic } = fioDomain
    const fioWallet = fioWallets.find((fioWallet: EdgeCurrencyWallet) => fioWallet.id === walletId)
    if (fioWallet == null) return
    navigation.navigate('fioDomainSettings', {
      fioWallet,
      fioDomainName: name,
      expiration,
      isPublic
    })
  }

  render() {
    const { fioAddresses, fioDomains, loading, navigation, theme } = this.props
    const { initLoading } = this.state
    const styles = getStyles(theme)

    const noFioDomainsText = `${lstrings.no} ${lstrings.title_fio_domains}`
    const noFioAddressesText = `${lstrings.no} ${lstrings.title_fio_address}`
    return (
      <>
        <SceneWrapper background="theme">
          <ScrollView style={styles.row}>
            <SceneHeader title={lstrings.title_fio_address} underline />
            <View style={styles.list}>
              {!fioAddresses.length && <EdgeText style={styles.noNames}>{noFioAddressesText}</EdgeText>}
              {fioAddresses.map((address: FioAddress) => (
                <FioNameRow
                  key={`${address.name}`}
                  name={address.name}
                  // @ts-expect-error
                  bundledTxs={address.bundledTxs}
                  icon={<Image source={fioAddressLogo} style={styles.iconImg} />}
                  theme={theme}
                  onPress={() => this.onAddressPress(address)}
                />
              ))}
            </View>
            <SceneHeader title={lstrings.title_fio_domains} underline withTopMargin />
            <View style={styles.list}>
              {!fioDomains.length && <EdgeText style={styles.noNames}>{noFioDomainsText}</EdgeText>}
              {fioDomains.map((domain: FioDomain) => (
                <FioNameRow
                  key={`${domain.name}`}
                  name={domain.name}
                  expiration={domain.expiration}
                  icon={<IonIcon name="ios-at" style={styles.iconIon} color={theme.icon} size={theme.rem(1.5)} />}
                  // @ts-expect-error
                  theme={theme}
                  onPress={() => this.onDomainPress(domain)}
                />
              ))}
            </View>
            <Fade visible={loading && !initLoading}>
              <ActivityIndicator color={theme.iconTappable} style={styles.loading} size="large" />
            </Fade>
          </ScrollView>

          <View>
            <ClickableText marginRem={[1, 1, 0]} onPress={() => navigation.navigate('fioAddressRegister', {})}>
              <View style={styles.actionButton}>
                <Fontello name="register-new-fio-icon" style={styles.actionIcon} color={theme.iconTappable} size={theme.rem(1)} />
                <EdgeText style={styles.buttonText}>{lstrings.fio_address_list_screen_button_register}</EdgeText>
              </View>
            </ClickableText>
            <ClickableText marginRem={[0, 1, 2, 1]} onPress={() => navigation.navigate('fioDomainRegister', {})}>
              <View style={styles.actionButton}>
                <Fontello name="register-custom-fio" style={styles.actionIcon} color={theme.iconTappable} size={theme.rem(1)} />
                <EdgeText style={styles.buttonText}>{lstrings.fio_address_list_domain_register}</EdgeText>
              </View>
            </ClickableText>
          </View>
        </SceneWrapper>

        {!initLoading ? null : <FullScreenLoader indicatorStyles={styles.loading} />}
      </>
    )
  }
}

const getStyles = cacheStyles((theme: Theme) => ({
  list: {
    display: 'flex',
    flexDirection: 'column',
    paddingTop: theme.rem(0.5)
  },
  loading: {
    flex: 1,
    marginTop: theme.rem(2.5),
    alignSelf: 'center'
  },
  row: {
    flex: 1
  },
  noNames: {
    color: theme.deactivatedText,
    fontSize: theme.rem(1),
    textAlign: 'center',
    padding: theme.rem(1)
  },
  buttonText: {
    marginLeft: theme.rem(0.5),
    color: theme.textLink,
    textAlign: 'center'
  },
  iconImg: {
    height: theme.rem(2.25),
    marginRight: theme.rem(1.5)
  },
  iconIon: {
    width: theme.rem(1.5),
    marginRight: theme.rem(1),
    textAlign: 'center'
  },
  actionButton: {
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center'
  },
  actionIcon: {
    marginTop: theme.rem(0.25)
  }
}))

export const FioAddressListScene = connect<StateProps, DispatchProps, OwnProps>(
  state => ({
    fioAddresses: state.ui.fioAddress.fioAddresses,
    fioDomains: state.ui.fioAddress.fioDomains,
    fioWallets: state.ui.wallets.fioWallets,
    loading: state.ui.fioAddress.fioAddressesLoading,
    isConnected: state.network.isConnected
  }),
  dispatch => ({
    async refreshAllFioAddresses() {
      await dispatch(refreshAllFioAddresses())
    }
  })
)(withTheme(FioAddressList))
