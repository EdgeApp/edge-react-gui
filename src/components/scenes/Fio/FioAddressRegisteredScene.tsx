import * as React from 'react'
import { Image, View } from 'react-native'

import { formatDate } from '../../../locales/intl'
import { lstrings } from '../../../locales/strings'
import { EdgeSceneProps } from '../../../types/routerTypes'
import { SceneWrapper } from '../../common/SceneWrapper'
import { FormattedText } from '../../legacy/FormattedText/FormattedText.ui'
import { cacheStyles, Theme, ThemeProps, withTheme } from '../../services/ThemeContext'
import { MainButton } from '../../themed/MainButton'

interface OwnProps extends EdgeSceneProps<'fioAddressRegisterSuccess'> {}

type Props = OwnProps & ThemeProps

export class FioAddressRegistered extends React.Component<Props> {
  renderExpDate() {
    const { theme, route } = this.props
    const { expiration } = route.params
    if (expiration != null) {
      const styles = getStyles(theme)

      return (
        <FormattedText style={styles.text}>
          {`${lstrings.fio_address_details_screen_expires} `}
          {formatDate(new Date(expiration))}
        </FormattedText>
      )
    }

    return null
  }

  render() {
    const { theme, navigation, route } = this.props
    const { fioName } = route.params
    const styles = getStyles(theme)

    return (
      <SceneWrapper background="theme">
        <View style={styles.view}>
          <View style={styles.texts}>
            <View style={styles.image}>
              <Image source={theme.fioAddressLogo} />
            </View>
            <FormattedText style={styles.text}>{lstrings.fio_address_details_screen_registered}</FormattedText>
            <FormattedText style={styles.title}>{fioName}</FormattedText>
            {this.renderExpDate()}
          </View>
          <MainButton marginRem={2} onPress={() => navigation.navigate('fioAddressList', {})} label={lstrings.title_fio_names} />
        </View>
      </SceneWrapper>
    )
  }
}

const getStyles = cacheStyles((theme: Theme) => ({
  view: {
    flex: 2,
    flexDirection: 'column',
    justifyContent: 'space-between',
    paddingHorizontal: theme.rem(1)
  },
  texts: {
    flex: 1,
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
    marginBottom: theme.rem(2.5)
  },
  text: {
    color: theme.primaryText,
    fontSize: theme.rem(1)
  },
  image: {
    marginLeft: theme.rem(0.25),
    marginBottom: theme.rem(3)
  },
  title: {
    fontSize: theme.rem(1.75),
    color: theme.primaryText,
    marginTop: theme.rem(1.25),
    marginBottom: theme.rem(0.75)
  },
  expiration: {
    fontSize: theme.rem(0.75),
    color: theme.primaryText,
    textAlign: 'center',
    marginTop: theme.rem(-0.5),
    paddingBottom: theme.rem(0.75)
  }
}))

export const FioAddressRegisteredScene = withTheme(FioAddressRegistered)
