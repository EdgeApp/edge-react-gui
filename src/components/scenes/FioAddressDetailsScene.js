// @flow

import type { EdgeCurrencyWallet } from 'edge-core-js'
import * as React from 'react'
import { Alert } from 'react-native'
import IonIcon from 'react-native-vector-icons/Ionicons'

import s from '../../locales/strings.js'
import { ConnectWalletsConnector as ConnectWallets } from '../../modules/FioAddress/components/ConnectWallets'
import { findWalletByFioAddress } from '../../modules/FioAddress/util'
import { connect } from '../../types/reactRedux.js'
import { type NavigationProp, type RouteProp } from '../../types/routerTypes.js'
import { SceneWrapper } from '../common/SceneWrapper'
import { showError } from '../services/AirshipInstance'
import { type Theme, type ThemeProps, cacheStyles, withTheme } from '../services/ThemeContext'
import { SettingsHeaderRow } from '../themed/SettingsHeaderRow'
import { SettingsTappableRow } from '../themed/SettingsTappableRow.js'

type StateProps = {
  fioWallets: EdgeCurrencyWallet[]
}

type OwnProps = {
  navigation: NavigationProp<'fioAddressDetails'>,
  route: RouteProp<'fioAddressDetails'>
}

type LocalState = {
  fioWalletLoading: boolean,
  fioWallet: EdgeCurrencyWallet | null
}

type Props = StateProps & ThemeProps & OwnProps

class FioAddressDetails extends React.Component<Props, LocalState> {
  state: LocalState = {
    fioWalletLoading: false,
    fioWallet: null
  }

  componentDidMount() {
    const { route } = this.props
    const { fioAddressName } = route.params
    if (!fioAddressName) {
      Alert.alert(s.strings.fio_address_details_screen_alert_title, s.strings.fio_address_details_screen_alert_message, [
        { text: s.strings.fio_address_details_screen_alert_button }
      ])
    }
    this.findFioWallet()
  }

  findFioWallet = async () => {
    const { fioWallets, route } = this.props
    const { fioAddressName } = route.params

    this.setState({ fioWalletLoading: true })
    const fioWallet = await findWalletByFioAddress(fioWallets, fioAddressName)
    this.setState({ fioWalletLoading: false, fioWallet })
  }

  _onPressAccountSettings = (): void => {
    const { navigation, route } = this.props
    const { fioAddressName } = route.params
    const { fioWallet } = this.state
    if (fioWallet) {
      navigation.navigate('fioAddressSettings', {
        fioWallet,
        fioAddressName,
        refreshAfterRenew: true
      })
    } else {
      showError(s.strings.fio_wallet_missing_for_fio_address)
    }
  }

  renderAccountSettings = () => {
    const { theme } = this.props
    const styles = getStyles(theme)

    return (
      <SettingsTappableRow label={s.strings.fio_address_details_screen_manage_account_settings} onPress={this._onPressAccountSettings}>
        <IonIcon name="ios-settings" color={theme.icon} style={styles.settingsIcon} />
      </SettingsTappableRow>
    )
  }

  render() {
    const { theme, route } = this.props
    const { fioAddressName } = route.params

    return (
      <SceneWrapper background="header">
        {this.renderAccountSettings()}
        <SettingsHeaderRow
          icon={<IonIcon name="ios-link" color={theme.primaryText} size={theme.rem(1.5)} />}
          label={s.strings.fio_address_details_connect_to_wallets}
        />
        <ConnectWallets fioAddressName={fioAddressName} fioWallet={this.state.fioWallet} disabled={this.state.fioWalletLoading} />
      </SceneWrapper>
    )
  }
}

const getStyles = cacheStyles((theme: Theme) => ({
  text: {
    color: theme.primaryText,
    fontSize: theme.rem(1)
  },
  settingsIcon: {
    fontSize: theme.rem(1.5),
    paddingHorizontal: theme.rem(0.5)
  },
  settingsWarning: {
    fontSize: theme.rem(1),
    color: theme.warningText,
    paddingHorizontal: theme.rem(0.5)
  }
}))

export const FioAddressDetailsScene = connect<StateProps, {}, OwnProps>(
  state => ({
    fioWallets: state.ui.wallets.fioWallets
  }),
  dispatch => ({})
)(withTheme(FioAddressDetails))
