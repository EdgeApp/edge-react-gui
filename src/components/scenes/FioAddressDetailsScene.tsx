import { EdgeCurrencyWallet } from 'edge-core-js'
import * as React from 'react'
import { Alert } from 'react-native'
import IonIcon from 'react-native-vector-icons/Ionicons'

import s from '../../locales/strings'
import { ConnectWalletsConnector as ConnectWallets } from '../../modules/FioAddress/components/ConnectWallets'
import { BUNDLED_TXS_AMOUNT_ALERT, findWalletByFioAddress } from '../../modules/FioAddress/util'
import { connect } from '../../types/reactRedux'
import { NavigationProp, RouteProp } from '../../types/routerTypes'
import { SceneWrapper } from '../common/SceneWrapper'
import { showError } from '../services/AirshipInstance'
import { cacheStyles, Theme, ThemeProps, withTheme } from '../services/ThemeContext'
import { EdgeText } from '../themed/EdgeText'
import { SettingsHeaderRow } from '../themed/SettingsHeaderRow'
import { SettingsTappableRow } from '../themed/SettingsTappableRow'

type StateProps = {
  fioWallets: EdgeCurrencyWallet[]
}

type OwnProps = {
  navigation: NavigationProp<'fioAddressDetails'>
  route: RouteProp<'fioAddressDetails'>
}

type LocalState = {
  fioWalletLoading: boolean
  fioWallet: EdgeCurrencyWallet | null
}

type Props = StateProps & ThemeProps & OwnProps

export class FioAddressDetails extends React.Component<Props, LocalState> {
  state: LocalState = {
    fioWalletLoading: false,
    fioWallet: null
  }

  componentDidMount() {
    const { route } = this.props
    const { fioAddressName } = route.params
    if (!fioAddressName) {
      Alert.alert(s.strings.fio_address_details_screen_alert_title, s.strings.fio_address_details_screen_alert_message, [
        { text: s.strings.fio_address_details_screen_alert_button }
      ])
    }
    this.findFioWallet()
  }

  findFioWallet = async () => {
    const { fioWallets, route } = this.props
    const { fioAddressName } = route.params

    this.setState({ fioWalletLoading: true })
    const fioWallet = await findWalletByFioAddress(fioWallets, fioAddressName)
    this.setState({ fioWalletLoading: false, fioWallet })
  }

  _onPressAccountSettings = (): void => {
    const { navigation, route } = this.props
    const { fioAddressName, bundledTxs } = route.params
    const { fioWallet } = this.state
    if (fioWallet) {
      navigation.navigate('fioAddressSettings', {
        fioWallet,
        fioAddressName,
        bundledTxs,
        refreshAfterAddBundledTxs: true
      })
    } else {
      showError(s.strings.fio_wallet_missing_for_fio_address)
    }
  }

  renderAccountSettings = () => {
    const {
      theme,
      route: {
        params: { bundledTxs }
      }
    } = this.props
    const styles = getStyles(theme)

    if (bundledTxs < BUNDLED_TXS_AMOUNT_ALERT) {
      return (
        <SettingsTappableRow onPress={this._onPressAccountSettings}>
          <IonIcon name="ios-warning" color={theme.warningIcon} style={styles.settingsIcon} />
          <EdgeText style={styles.settingsWarning} numberOfLines={4}>
            {!bundledTxs ? s.strings.fio_address_details_no_bundled_txs : s.strings.fio_address_details_bundled_txs_out_soon}
          </EdgeText>
        </SettingsTappableRow>
      )
    }

    return (
      <SettingsTappableRow label={s.strings.fio_address_details_screen_manage_account_settings} onPress={this._onPressAccountSettings}>
        <IonIcon name="ios-settings" color={theme.icon} style={styles.settingsIcon} />
      </SettingsTappableRow>
    )
  }

  render() {
    const { theme, route } = this.props
    const { fioAddressName, bundledTxs } = route.params
    const styles = getStyles(theme)
    const bundledTxsLabel = `${s.strings.fio_address_details_screen_bundled_txs}: ${bundledTxs}`

    return (
      <SceneWrapper background="header">
        <EdgeText style={styles.bundledTxs}>{bundledTxsLabel}</EdgeText>
        {this.renderAccountSettings()}
        <SettingsHeaderRow
          icon={<IonIcon name="ios-link" color={theme.primaryText} size={theme.rem(1.5)} />}
          label={s.strings.fio_address_details_connect_to_wallets}
        />
        <ConnectWallets fioAddressName={fioAddressName} fioWallet={this.state.fioWallet} disabled={this.state.fioWalletLoading} />
      </SceneWrapper>
    )
  }
}

const getStyles = cacheStyles((theme: Theme) => ({
  text: {
    color: theme.primaryText,
    fontSize: theme.rem(1)
  },
  bundledTxs: {
    fontSize: theme.rem(0.75),
    color: theme.primaryText,
    textAlign: 'center',
    marginTop: theme.rem(-0.5),
    paddingBottom: theme.rem(0.75)
  },
  settingsIcon: {
    fontSize: theme.rem(1.5),
    paddingHorizontal: theme.rem(0.5)
  },
  settingsWarning: {
    fontSize: theme.rem(0.75),
    color: theme.warningText,
    paddingLeft: theme.rem(0.5),
    width: '75%'
  }
}))

export const FioAddressDetailsScene = connect<StateProps, {}, OwnProps>(
  state => ({
    fioWallets: state.ui.wallets.fioWallets
  }),
  dispatch => ({})
)(withTheme(FioAddressDetails))
