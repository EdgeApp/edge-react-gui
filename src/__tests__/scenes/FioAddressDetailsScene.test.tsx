import { describe, expect, it } from '@jest/globals'
import * as React from 'react'
import { createRenderer } from 'react-test-renderer/shallow'

import { FioAddressDetails } from '../../components/scenes/Fio/FioAddressDetailsScene'
import { getTheme } from '../../components/services/ThemeContext'
import { fakeSceneProps } from '../../util/fake/fakeSceneProps'

describe('FioAddressDetails', () => {
  it('should render with loading props', () => {
    const renderer = createRenderer()

    const fakeWallet: any = {
      blockHeight: 34,
      currencyCode: 'FIO',
      date: 220322,
      nativeAmount: '100',
      networkFee: '1',
      ourReceiveAddresses: ['FioAddress'],
      signedTx: '0xdgs3442',
      stakingStatus: { stakedAmounts: [] },
      txid: '0x34346463'
    }

    const actual = renderer.render(
      <FioAddressDetails
        {...fakeSceneProps('fioAddressDetails', {
          fioAddressName: 'Fio@edge',
          bundledTxs: 100
        })}
        fioWallets={[fakeWallet]}
        theme={getTheme()}
      />
    )

    expect(actual).toMatchSnapshot()
  })
})
