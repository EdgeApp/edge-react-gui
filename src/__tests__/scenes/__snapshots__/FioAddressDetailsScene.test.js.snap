// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FioAddressDetails should render with loading props 1`] = `
<SceneWrapper
  background="header"
>
  <WithTheme(EdgeTextComponent)
    style={
      Object {
        "color": "#FFFFFF",
        "fontSize": 17,
        "marginTop": -11,
        "paddingBottom": 17,
        "textAlign": "center",
      }
    }
  >
    Remaining bundled transactions: 100
  </WithTheme(EdgeTextComponent)>
  <ForwardRef
    label="Crypto Handle Settings"
    onPress={[Function]}
  >
    <Icon
      allowFontScaling={false}
      color="#FFFFFF"
      name="ios-settings"
      size={12}
      style={
        Object {
          "fontSize": 34,
          "paddingHorizontal": 11,
        }
      }
    />
  </ForwardRef>
  <WithTheme(SettingsHeaderRowComponent)
    icon={
      <Icon
        allowFontScaling={false}
        color="#FFFFFF"
        name="ios-link"
        size={34}
      />
    }
    label="Choose wallets to connect to your FIO Crypto Handle"
  />
  <Memo(Connect(WithTheme(ConnectWallets)))
    disabled={false}
    fioAddressName="Fio@edge"
    fioWallet={null}
  />
</SceneWrapper>
`;
